internal/database/database.go

```diff
@@ -0,0 +1,287 @@
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Iterate rows
+	for rows.Next() {
+
+		// Read rows into struct
+		fi := &FileInfo{}
+		err := rows.Scan(&fi.FileID, &fi.DatasetID, &fi.DisplayFileName, &fi.FileName, &fi.FileSize,
+			&fi.DecryptedFileSize, &fi.DecryptedFileChecksum, &fi.DecryptedFileChecksumType, &fi.Status)
+		if err != nil {
+			log.Error(err)
+			return nil, err
+		}
+
+		// local_ega_ebi.file:file_size is actually the size of the archive file without header
+		// so we need to increase the encrypted file size by the length of the header if the user
+		// downloaded the files in encrypted format. I set it as 124 which seems to be the default
+		// length, but if files can have greater headers, then we can calculate the length with
+		// fd := GetFile() --> len(fd.Header)
+		fi.FileSize = fi.FileSize + 124
+
+		// Add structs to array
+		files = append(files, fi)
+	}
+
+	return files, nil
+}
+
+// CheckDataset checks if dataset name exists
+func (dbs *SQLdb) CheckDataset(dataset string) (bool, error) {
+	var (
+		r     bool  = false
+		err   error = nil
+		count int   = 0
+	)
+
+	for count == 0 || (err != nil && count < dbRetryTimes) {
+		r, err = dbs.checkDataset(dataset)
+		count++
+	}
+	return r, err
+}
+
+// checkDataset is the actual function performing work for CheckDataset
+func (dbs *SQLdb) checkDataset(dataset string) (bool, error) {
+	dbs.checkAndReconnectIfNeeded()
+
+	db := dbs.DB
+	const query = "SELECT DISTINCT dataset_stable_id FROM local_ega_ebi.filedataset WHERE dataset_stable_id = $1"
+
+	var datasetName string
+	if err := db.QueryRow(query, dataset).Scan(&datasetName); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// CheckFilePermission checks if user has permissions to access the dataset the file is a part of
+func (dbs *SQLdb) CheckFilePermission(fileID string) (string, error) {
+	var (
+		r     string = ""
+		err   error  = nil
+		count int    = 0
+	)
+
+	for count == 0 || (err != nil && count < dbRetryTimes) {
+		r, err = dbs.checkFilePermission(fileID)
+		count++
+	}
+	return r, err
+}
+
+// checkFilePermission is the actual function performing work for CheckFilePermission
+func (dbs *SQLdb) checkFilePermission(fileID string) (string, error) {
+	dbs.checkAndReconnectIfNeeded()
+
+	db := dbs.DB
+	const query = "SELECT dataset_id FROM local_ega_ebi.file_dataset WHERE file_id = $1"
+
+	var datasetName string
+	if err := db.QueryRow(query, fileID).Scan(&datasetName); err != nil {
+		return "", err
+	}
+
+	return datasetName, nil
+}
+
+// FileDownload details are used for downloading a file
+type FileDownload struct {
+	ArchivePath string
+	ArchiveSize int
+	Header      []byte
+}
+
+// GetFile retrieves the file header
+func (dbs *SQLdb) GetFile(fileID string) (*FileDownload, error) {
+	var (
+		r     *FileDownload = nil
+		err   error         = nil
+		count int           = 0
+	)
+
+	for count == 0 || (err != nil && count < dbRetryTimes) {
+		r, err = dbs.getFile(fileID)
+		count++
+	}
```

Comment:
This is not trivial to read and could possibly be separated in a for loop with a nested if.

Reply:
This logic was copied `sda-pipeline` from https://github.com/neicnordic/sda-pipeline/blob/master/internal/database/db.go
But you're right, maybe we could make it a tad easier on the eyes by breaking it down to parts:
Fixed with https://github.com/neicnordic/sda-download/pull/15/commits/9511baf9218c955e1222cdec5949922093385103